### PR TITLE
Bin optimizer

### PR DESCRIPTION
**Pull request by domoritz**
@Saba9 Please continue

### CODE REVIEW

**Comment by domoritz** [comment]
When this is ready, please update the title and labels.

**Comment by domoritz** [comment]
Please rebase to master. this should be easy now that we sue Vega 5 and the new SignalRefWrapper.

File: src/compile/data/assemble.ts [hunk 1]
```diff
@@ -203,8 +203,8 @@ export function assembleFacetData(root: FacetNode): VgData[] {
 export function assembleRootData(dataComponent: DataComponent, datasets: Dict<InlineDataset>): VgData[] {
   const data: VgData[] = [];
 
-  // roots.forEach(debug);
-  // draw(roots);
+  // dataComponent.sources.forEach(debug);
+  // draw(dataComponent.sources);
```
Comment (domoritz) [inline src/compile/data/assemble.ts]: Thanks

File: test/compile/scale/domain.test.ts [hunk 2]
```diff
@@ -125,12 +126,8 @@ describe('compile/scale', () => {
               y: fieldDef
             }
           });
-
-          expect(testParseDomainForChannel(model, 'y')).toEqual([
-            {
-              signal: '[bin_maxbins_15_origin_bins.start, bin_maxbins_15_origin_bins.stop]'
-            }
-          ]);
+          const domain = testParseDomainForChannel(model, 'y')[0] as SignalRefWrapper;
```
Comment (domoritz) [inline test/compile/scale/domain.test.ts]:
```suggestion
          const domain = testParseDomainForChannel(model, 'y')[0] as SignalRef;
```

File: src/compile/data/optimizers.ts [hunk 3]
```diff
@@ -343,3 +350,51 @@ export class MergeAggregateNodes extends BottomUpOptimizer {
     return this.flags;
   }
 }
+
+/**
+ * Merge bin nodes and move bin nodes up through forks but stop at filters and parse.
```
Comment (domoritz) [inline src/compile/data/optimizers.ts]:
```suggestion
 * Merge bin nodes and move bin nodes up through forks but stop at filters and parse as we want them to stay before the bin node.
```

File: src/compile/scale/range.ts [hunk 4]
```diff
@@ -116,10 +116,13 @@ function getRangeStep(model: UnitModel, channel: 'x' | 'y'): number | SignalRef
...
       const sizeType = getSizeType(channel);
       const sizeSignal = model.getName(sizeType);
-      return new SignalRefWrapper(() => `${model.getSignalName(sizeSignal)} / (${model.getSignalName(binCount)})`);
+      return new SignalRefWrapper(() => {
+        const signalName = model.getSignalName(binSignal);
+        const binCount = `(${signalName}.stop - ${signalName}.start) / ${signalName}.step`;
+        return `${model.getSignalName(sizeSignal)} / (${binCount})`;
+      });
     }
     // TODO: handle binned case
```
Comment (domoritz) [inline src/compile/scale/range.ts]: Don't we do this above?

File: src/compile/scale/range.ts [hunk 5]
```diff
@@ -116,10 +116,13 @@ function getRangeStep(model: UnitModel, channel: 'x' | 'y'): number | SignalRef
       const binSignal = model.getName(vgField(fieldDef, {suffix: 'bins'}));
 
       // TODO: extract this to be range step signal
-      const binCount = `(${binSignal}.stop - ${binSignal}.start) / ${binSignal}.step`;
       const sizeType = getSizeType(channel);
       const sizeSignal = model.getName(sizeType);
```
Comment (domoritz) [inline src/compile/scale/range.ts]: Just wondering, why do we need this at all? What happens if we delete all of it and never generate a rangeStep for binned fields?

Comment (domoritz) [reply src/compile/scale/range.ts]: We may need it for ordinal data so try that.

**Comment by domoritz** [comment]
Probably needs to merge master again.

File: src/compile/scale/range.ts (hunk 5 above)
Comment (Saba9) [reply src/compile/scale/range.ts]: Ahh, looks like we do need the rangeStep signal if you look at the updated examples.

**Comment by domoritz** [comment]
Is this ready?

**Comment by domoritz** [comment]
There is some untested code that we should cover in tests: https://codecov.io/gh/vega/vega-lite/compare/2307cc3ee83e4f5bf80a9106941964ffaa58d5b2...3d2fb4ed0b9a300bb1319adb5250ad2189f76a1d/src/src/compile/scale/range.ts#L317

**Comment by domoritz** [comment]
Ping @Saba9

**Comment by Saba9** [comment]
yeah I believe this is ready. Just need to write some tests for coverage to pass

**Comment by Saba9** [comment]
actually I need to revert a change

**Comment by Saba9** [comment]
Alright, should be ready for review. I can squash the commits before we merge.

**Comment by domoritz** [comment]
Thank you!

**Comment by domoritz** [comment]
I think something is missing here. We are merging only whole bin nodes but not individual bin operations. Am I right @Saba9?

**Pull request merged**